virtual-display-driver-cli/src/client.rs

```diff
@@ -1,110 +1,139 @@
+            eyre::bail!("received unexpected reply from driver pipe");
+        };
+
+        Ok(Self { writer, state })
     }
 
-    pub fn list(&mut self) -> eyre::Result<Vec<Monitor>> {
-        self.send_command(&driver_ipc::Command::RequestState)?;
-        let response = self.receive_command()?;
-
-        match response {
-            driver_ipc::Command::ReplyState(state) => Ok(state),
-            _ => eyre::bail!("received unexpected reply from driver pipe"),
-        }
+    pub fn monitors(&self) -> &[Monitor] {
+        &self.state
     }
 
     pub fn get(&mut self, id: driver_ipc::Id) -> eyre::Result<Monitor> {
-        let monitors = self.list()?;
-        let monitor = monitors.into_iter().find(|monitor| monitor.id == id);
+        let monitor = self.state.iter().find(|monitor| monitor.id == id);
 
         match monitor {
-            Some(monitor) => Ok(monitor),
-            None => eyre::bail!("no virtual monitor with ID {} found", id),
+            Some(monitor) => Ok(monitor.clone()),
+            None => eyre::bail!("virtual monitor with ID {} not found", id),
         }
     }
 
     pub fn notify(&mut self, monitors: Vec<driver_ipc::Monitor>) -> eyre::Result<()> {
         let command = driver_ipc::Command::DriverNotify(monitors);
 
-        self.send_command(&command)?;
+        send_command(&mut self.writer, &command)?;
 
         Ok(())
     }
 
+    pub fn validate_has_ids(&self, ids: &[driver_ipc::Id]) -> eyre::Result<()> {
+        let ids = ids.iter().copied().collect::<BTreeSet<_>>();
+        let existing_ids = self
+            .state
+            .iter()
+            .map(|monitor| monitor.id)
+            .collect::<BTreeSet<_>>();
+        let missing_ids = ids.difference(&existing_ids).copied().collect::<Vec<_>>();
+
+        if missing_ids.is_empty() {
+            Ok(())
+        } else if missing_ids.len() == 1 {
+            eyre::bail!("virtual monitor with ID {} not found", missing_ids[0])
+        } else {
+            eyre::bail!(
+                "virtual monitors with IDs not found: {}",
+                missing_ids.join_with(", ")
+            )
+        }
+    }
+
     pub fn remove(&mut self, ids: Vec<driver_ipc::Id>) -> eyre::Result<()> {
         let command = driver_ipc::Command::DriverRemove(ids);
 
-        self.send_command(&command)?;
+        send_command(&mut self.writer, &command)?;
 
         Ok(())
     }
 
     pub fn remove_all(&mut self) -> eyre::Result<()> {
         let command = driver_ipc::Command::DriverRemoveAll;
 
-        self.send_command(&command)?;
+        send_command(&mut self.writer, &command)?;
 
         Ok(())
     }
 
     pub fn new_id(&mut self, preferred_id: Option<driver_ipc::Id>) -> eyre::Result<driver_ipc::Id> {
-        let monitors = self.list()?;
+        let existing_ids = self
+            .state
+            .iter()
+            .map(|monitor| monitor.id)
+            .collect::<HashSet<_>>();
```

Comment:
Implementation note: I know it's a little silly to construct a `HashSet` just to tell if there's an existing ID or not, especially considering that there'll probably only ever be ~10 monitors (maybe 1-2 dozen if there's a workaround for the hard limit at the moment).

I did do some quick testing though, and collecting into a `HashSet` seemed to be very fast for even a lot of items, and it makes finding the next-smallest ID extremely quick. As a bonus, I felt the code was a little cleaner rather than needing to linearly search the list on each iteration.

Reply:
I actually was profiling inserting into a hashset yesterday (at the worst case mine held about 400 items). It took, at worst, about 40μs, so performance in our use-case is easily a non-issue.